onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc

```diff
@@ -38,51 +42,252 @@ class ResizeOpBuilder : public BaseOpBuilder {
+  const std::string resize_mode = GetOnnxAttr(node_helper, onnx_mode_attr);
+  ORT_RETURN_IF("cubic" == resize_mode, "Resize doesn't support cubic mode!");
+
+  const std::string coordinate_mode = GetOnnxAttr(node_helper, onnx_coord_transf_mode_attr);
+  if ((coordinate_mode != "half_pixel") && (coordinate_mode != "align_corners")) {
+    std::string msg = coordinate_mode + " mode not supported, QNN Resize only supports align_corners and half_pixel";
     return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, msg);
   }
 
-  if ("nearest" == resize_mode) {
-    const std::string& nearest_mode = node_helper.Get("nearest_mode", "round_prefer_floor");
-    ORT_RETURN_IF_NOT("floor" == nearest_mode, "QNN Resize only support nearest_mode: floor!");
+  // Check for a valid "nearest_mode" if the mode is "nearest".
+  if (resize_mode == "nearest") {
+    // NOTE: QNN's ResizeNearestNeighbor operator does not have a way to specify rounding (i.e., "nearest_mode").
+    // The output of the QNN ResizeNearestNeighbor operator is not always equivalent to ONNX's Resize
+    // operator with any single specific "nearest_mode".
+    //
+    // For some input/output shapes, QNN's ResizeNearestNeighbor is equivalent to ONNX's Resize with "round_prefer_floor".
+    // For other shapes, QNN's ResizeNearestNeighbor is equivalent to ONNX Resize with "round_prefer_ceil".
+    //
+    // From unit tests, I've found a relationship between input/output shapes and the equivalent ONNX "nearest_mode".
+    // If the new and old spatial dimensions are evenly divisible, the "nearest_mode" is "round_prefer_floor".
+    // Otherwise, the "nearest_mode" is "round_prefer_ceil".
+    //
+    // This relationship is probably incomplete/wrong.
+    //
+    // TODO: Ask Qualcomm what the correct "nearest_mode" should be,
+    // OR use QNN's own Resize operator once it works on QnnCpu.
+    const std::string& nearest_mode = GetOnnxAttr(node_helper, onnx_nearest_mode_attr);
+    ORT_RETURN_IF_NOT("floor" == nearest_mode, "QNN Resize only supports nearest_mode: floor!");  // This is wrong!
   }
 
   auto& input_0 = node_unit.Inputs()[0];
   std::vector<uint32_t> input_shape;
-  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(input_0.node_arg, input_shape), "Cannot get shape");
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(input_0.node_arg, input_shape),
+                    "QNN EP: Cannot get input shape for Resize op");
   if (input_shape.size() != 4) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN Resize only support 2D!");
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN Resize only supports 4D!");
+  }
+
+  const auto& output_0 = node_unit.Outputs()[0];
+  std::vector<uint32_t> output_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(output_0.node_arg, output_shape),
+                    "QNN EP: Cannot get output shape for Resize op");
+  if (output_shape.size() != 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN Resize only supports 4D!");
   }
```

Comment:
use ORT_RETURN_IF

Reply:
merge it with if (input_shape.size() != 4)
ORT_RETURN_IF(input_shape.size() != 4 || output_shape.size() != 4, "QNN Resize only supports 4D!");

---
In reply to: [1160306095](https://github.com/microsoft/onnxruntime/pull/15390#discussion_r1160306095) [](ancestors = 1160306095)